encoding/template/delta_encode.go

```diff
@@ -0,0 +1,59 @@
+package template
+
+import (
+	"io"
+
+	bitstream "github.com/dgryski/go-bitstream"
+	"github.com/mauricelam/genny/generic"
+)
+
+// GenericValue represents a generic value type.
+type GenericValue generic.Type
+
+// ForwardValueIterator allows iterating over a stream of GenericValue.
+type ForwardValueIterator interface {
+	generic.Type
+	io.Closer
+	Next() bool
+	Err() error
+	Current() GenericValue
+	Rewind()
+}
+
+func encodeDeltaValue(
+	bitWriter *bitstream.BitWriter,
+	bitsPerEncodedValue int64,
+	valuesIt ForwardValueIterator,
+	subFn func(curr GenericValue, last GenericValue) int,
+) error {
+	// Encode the first value which is always a delta of 0.
+	if !valuesIt.Next() {
+		return valuesIt.Err()
+	}
+
+	// Write an extra bit to encode the sign of the delta.
+	if err := bitWriter.WriteBits(uint64(0), int(bitsPerEncodedValue)); err != nil {
```

Comment:
Hm shouldnt this be `bitWriter.WriteBits(valuesIt.Current(), bitsPerEncodedValue)` rather?

Reply:
The first value should be encoded as 0.